File: testdata/compile/gammar/stmt_base.go
package main

import "github.com/heyuuu/gophp/php/def"

func init() {
	f := def.NewFile("", false)

	f.TopFn("", func(d def.TopDefiner) def.Val {
		// empty stmt
		// expr stmt
		d.Int(1)
		// const stmt
		d.DeclConst("a", d.Int(1))
		// echo stmt
		d.Echo("abc")
		d.EchoVal(d.VarValue("a"))
		// global stmt
		d.DeclGlobal("a")
		d.DeclGlobalVal(d.VarValue("a"))
		// static var stmt
		d.DeclStatic("a", nil)
		d.DeclStatic("b", d.Int(1))
		// unset stmt
		d.Unset(d.Var("a"))

		// if stmt
		if d.IsTrue(d.Equal(d.VarValue("a"), d.Int(1))) {
			// branch if
			d.Echo("if")
		} else if d.IsTrue(d.Equal(d.VarValue("b"), d.Int(2))) {
			// branch elseif
			d.Echo("elseif")
		} else {
			// branch else
			d.Echo("else")
		}

		// switch stmt
		switch switchVar := d.Equal(d.VarValue("a"), d.Int(2)); {
		case d.IsTrue(d.Equal(switchVar, d.Int(1))):
			// branch case 1
			break
		case d.IsTrue(d.Equal(switchVar, d.Int(2))):
			// branch case 2
			fallthrough
		case d.IsTrue(d.Equal(switchVar, d.Int(2))):
			// branch case 2
			break
		default:
			// branch default
		}

		// for stmt
		for d.Assign(d.Var("i"), d.Int(0)); d.IsTrue(d.Smaller(d.VarValue("i"), d.Int(10))); d.PostInc(d.Var("i")) {
			// branch for
			d.Echo("for")
		}

		// foreach stmt
		for iter := d.ForeachIterator(d.VarValue("arr")); iter.Valid(); iter.Next() {
			// todo
			d.Echo("foreach")
		}

		// while stmt
	_break_0:
		for {
			for {
				goto _break_0
			}
		}

		// do-while stmt
		for {
			continue
			break
		}

		return nil
	})
}

File: testdata/compile/gammar/stmt_base.php
<?php

// empty stmt
;
// expr stmt
1;
// const stmt
const A = 1;

// echo stmt
echo "abc";
echo $a;

// global stmt
global $a;
global $$a;

// static var stmt
static $a, $b = 1;

// unset stmt
unset($a);

// if stmt
if ($a === 1) {
    // branch if
    echo 'if';
} elseif ($b == 2) {
    // branch elseif
    echo 'elseif';
} else {
    // branch else
    echo 'else';
}

// switch stmt
switch ($a === 2) {
    case 1:
        // branch case 1
        break;
    case 2:
        // branch case 2
        // fallthrough
    case 3:
        break;
    default:
        // branch default
}

// for stmt
for ($i = 0; $i < 10; $i++) {
    // branch for
    echo "for";
}

// foreach stmt
foreach ($arr as $key => $value) {
    // branch foreach
    echo "foreach";
}

// while stmt
while (true) {
    while (true) {
        break 2;
    }
}

// do-while stmt
do {
    continue;
    break;
} while (true);

// try-catch-finally stmt
try {
    // branch try
    echo 1;
} catch (\Exception $e) {
    // branch catch
    echo "catch";
} finally {
    // branch finally
    echo "finally";
}
